src/ecs/resources/input.rs

```diff
@@ -206,16 +243,174 @@ impl InputHandler {
         self.mouse_position
     }
 
-    /// Gets the change in position since this function was last called.
-    ///
-    /// This function will return (0, 0) when it cannot calculate a change, either for the reasons
-    /// given above or on the first call.
-    pub fn mouse_position_change(&mut self) -> (i32, i32) {
-        let out = match (self.mouse_position, self.previous_mouse_position) {
+    /// Gets the change in position since the last frame.
+    pub fn mouse_position_change(&self) -> (i32, i32) {
+        match (self.mouse_position, self.previous_mouse_position) {
             (Some(current), Some(previous)) => (current.0 - previous.0, current.1 - previous.1),
             _ => (0, 0),
-        };
-        self.previous_mouse_position = self.mouse_position;
-        out
+        }
+    }
+
+    /// Returns an iterator over the buttons that are currently pressed
+    pub fn pressed_buttons(&self) -> PressedButtons {
+        let mouse_button_convert = mouse_button_to_button as fn(&MouseButton) -> Button;
+        let key_convert = key_to_button as fn(&VirtualKeyCode) -> Button;
+        let mouse_buttons = self.pressed_mouse_buttons().map(mouse_button_convert);
+        let keys = self.pressed_keys().map(key_convert);
+        PressedButtons { iterator: mouse_buttons.chain(keys) }
+    }
+
+    /// Checks if the given button is currently pressed.
+    pub fn button_pressed(&self, button: Button) -> bool {
+        match button {
+            Button::Key(k) => self.key_down(k),
+            Button::Mouse(b) => self.mouse_button_down(b),
+        }
     }
+
+    /// Checks if all the given buttons are pressed.
+    pub fn buttons_pressed(&self, buttons: &[Button]) -> bool {
+        buttons.iter().all(|b| self.button_down(*b))
+    }
+
+    /// Checks if the given mouse button was pressed on this frame.
+    pub fn button_down(&self, button: Button) -> bool {
+        match button {
+            Button::Key(k) => self.key_down(k),
+            Button::Mouse(b) => self.mouse_button_down(b),
+        }
+    }
+
+    /// Checks if the all the given buttons are being pressed and at least one was pressed this frame.
+    pub fn buttons_down(&self, buttons: &[Button]) -> bool {
+        buttons.iter().any(|&b| self.button_down(b)) &&
+        buttons.iter().all(|&b| self.button_pressed(b))
+    }
+
+    /// Returns the value of an axis by the i32 id, if the id doesn't exist this returns None.
+    pub fn axis_value(&self, id: i32) -> Option<f32> {
+        if let Some(a) = self.axes.get(&id) {
+            let pos = self.button_down(a.pos);
+            let neg = self.button_down(a.neg);
+            if pos == neg {
```

Comment:
```Rust
Some(if pos == neg {
                     0.0
                 } else if pos {
                1.0
            } else {
                -1.0
            })
```
This is the rustfmt version of that.  In my opinion it really hurts readability.